DOC new example for time-series forecasting with lagged features and prediction intervals

rprkh opened the pull request:
#### Reference Issues/PRs
Fixes #24378

#### What does this implement/fix? Explain your changes.
Includes an example for time-series forecasting that makes use of pandas-engineered lagged features and `HistGradientBoostingRegressor` with prediction intervals on the Bike Sharing Demand dataset.

Original Tutorial Code: https://nbviewer.org/github/ogrisel/euroscipy-2022-time-series/blob/main/plot_time_series_feature_engineering.ipynb
Trimmed Version: https://colab.research.google.com/drive/1mhvqNuXXBCm9Cp_wHulJ4mIHPEqOuo1s#scrollTo=PXyLv6esVTMj

#### Any other comments?
/cc @ogrisel

rprkh committed:
new time series example

rprkh committed:
remove unused imports and ignore warnings

rprkh committed:
lint

rprkh committed:
Merge branch 'main' into doc_new_example_time_series

ArturoAmorQ commented on examples/applications/plot_time_series_lagged_features.py:
```diff
@@ -0,0 +1,433 @@
+plt.subplot(224)
+create_boxplot("windspeed", "red")
+
+# %%
+# Exploring the hourly count of bikes over weekdays:
+plt.figure(figsize=(15, 10))
+sns.pointplot(x=df["hour"].astype(int), y=df["count"], hue=df["weekday"].astype(int))
+plt.title("Hourly Count for Weekdays", fontsize=20)
+plt.xlabel("Hour", fontsize=15)
+plt.ylabel("Count", fontsize=15)
+plt.xticks(fontsize=10)
+plt.yticks(fontsize=10)
+plt.show()
+
+# %%
+# Exploring the hourly count of bikes over different seasons:
+plt.figure(figsize=(15, 10))
+sns.pointplot(x=df["hour"].astype(int), y=df["count"], hue=df["season"])
+plt.title("Hourly Count for Seasons", fontsize=20)
+plt.xlabel("Hour", fontsize=15)
+plt.ylabel("Count", fontsize=15)
+plt.xticks(fontsize=10)
+plt.yticks(fontsize=10)
+plt.show()
```
This two plots are quite interesting to explore the data! Following the logic of referencing the [Time-related feature engineering example](https://scikit-learn.org/stable/auto_examples/applications/plot_cyclical_feature_engineering.html) for data exploration, I suggest you open a PR to rather add this plots on that notebook.

Just keep in mind that "weekdays" is not the same as "days of the week" :)

ArturoAmorQ commented on examples/applications/plot_time_series_lagged_features.py:
```diff
@@ -0,0 +1,433 @@
+    toc = perf_counter()
+    for key, value in cv_results.items():
+        if key.startswith("test_"):
+            print(f"{key[5:]}: {value.mean():.3f} ± {value.std():.3f}")
```
Instead of printing the results each time, maybe we can add the results to a dataframe in order to give a summary as a table similar to [the one on this example](https://scikit-learn.org/stable/auto_examples/ensemble/plot_gradient_boosting_quantile.html#analysis-of-the-error-metrics).

Please do let us know if you need help on this regard.

ogrisel replied:
In particular, this will allow to put the analysis where we compare the impact of the choice of the training loss for each model on the test set evaluation metric in a single place that would not make the reader scroll up and down across sections to compare the numbers for different models on a given metric.

ArturoAmorQ commented on examples/applications/plot_time_series_lagged_features.py:
```diff
@@ -0,0 +1,433 @@
+plt.show()
+
+# %%
+# Concluding Remarks
```
We can use a section conclusion and going further to just mention uncertainty estimation with `MAPIE` and recursive forecasting with `sktime`.

ArturoAmorQ reviewed (commented):
Thanks for the PR @rprkh. This will certainly be a nice addition to the example gallery! Here is batch of comments, I hope they are not too overwhelming and remember that in case you need help on our side, we are glad to help :)

ogrisel commented on examples/applications/plot_time_series_lagged_features.py:
```diff
@@ -0,0 +1,433 @@
+mean_absolute_percentage_error(y_test, y_pred)
+
+# %%
+# The error rate of this model is better than our naive shuffling
+# train-test split. This is quite expected but maybe the first split
```
Actually, it's not necessarily that the model is cheating but that when we use a shuffled train-test split, we are evaluating the model performance on an easier task (predicting on data points that are closer hence more similar to the training points) but that this measure does not reflect the ability of the model of making good predictions on data points collected in the future of the training period.

In other words, the generalization error measured via a shuffled trained test split is too optimistic. The generalization via a time-based split is likely to more representative of the true performance of the regression model if we were to deploy it for real.

ogrisel commented on examples/applications/plot_time_series_lagged_features.py:
```diff
@@ -0,0 +1,433 @@
+    "5th percentile",
+    "95th percentile",
+]
+for ax, pred, label in zip(axes, predictions, labels):
+    ax.scatter(pred, y_test.values, alpha=0.3, label=label)
+    ax.plot([0, y.max()], [0, y.max()], "--", label="Perfect model")
+    ax.set(
+        xlim=(0, y.max()),
+        ylim=(0, y.max()),
+        xlabel="Predicted demand",
+        ylabel="True demand",
+    )
+    ax.legend()
```
Could you try to use the residuals vs predicted plot instead (see the docstring of `PredictionErrorDisplay`). I think that might make it easier to analyze the behavior of each model.

rprkh replied:
Updated the plot to show "residual_vs_predicted" instead of "actual_vs_predicted"

ogrisel commented on examples/applications/plot_time_series_lagged_features.py:
```diff
@@ -0,0 +1,433 @@
+
+# %%
+# Looking at the performance of non-linear regression models vs
+# perfect models:
```
There are no "perfect models" directly displayed on those plots. For instance, the "perfect" models for the 5%-percentile estimation would have 95% of the points above the diagonal and 5% below. Furthermore, those fractions should stay uniform when binning horizontally. But it's hard to see if our models are good or not in that respect using those plots. Maybe they are maybe not.

ogrisel commented on examples/applications/plot_time_series_lagged_features.py:
```diff
@@ -0,0 +1,433 @@
+    label="GBRT 90% interval",
+)
+_ = ax.legend()
+
```
After each plot we need a paragraph to explain how to analyze what is displayed. Here it's interesting to notice that the blue area between the 5% and 95% percentile estimators has a width that varies with the time of the day:

- at night, the blue band is much narrower: the pair of models is quite certain that there will be a small number of bike rentals. And furthermore those seem correct in the sense that the actual demand stays in that blue band;
- during the day, the blue band is much wider: the uncertainty grows, probably because of the variability of the weather that can have a very large impact, especially on a week-end day;
- we can also see that during week-days, the commute pattern is still visible in the 5% and 95% estimations;
- finally, it is expected that 10% of the time, the actual demand does not lie between the 5% and 95% percentile estimates. On this test span, it the actual demand seems to be higher, especially during the rush ours. It might reveal that our 95% percentile estimator underestimates the demand peaks. This could be be quantitatively confirmed by computing empirical coverage numbers as done in this section of another example that we should link from this time-series forecasting example:

https://scikit-learn.org/stable/auto_examples/ensemble/plot_gradient_boosting_quantile.html#calibration-of-the-confidence-interval

ogrisel commented on examples/applications/plot_time_series_lagged_features.py:
```diff
@@ -0,0 +1,433 @@
+# Through this example we explored time series forecasting using lagged
+# features. We compared a naive regression (using the standardized
+# `train_test_split`) with a proper time series evaluation strategy using
+# `TimeSeriesSplit`. We observed that the model trained using `TimeSerieSplit`
+# gives a better Mean Average Percentage Error (MAPE) rate than the naive
+# method. We also analyzed the predictive uncertainty of our model via
```
Yes and in particular, it's not "better". The two MAPE numbers cannot be compared because they are not evaluated on the sample test sets.

It's the shuffle-split procedure that is not reflecting the true setting and as a result produces an overly optimistic MAPE number.

ogrisel commented on examples/applications/plot_time_series_lagged_features.py:
```diff
@@ -0,0 +1,433 @@
+# method. We also analyzed the predictive uncertainty of our model via
+# Quantile Regression. Predictions based on the 5th, 50th and 95th
+# percentile using `loss="quantile"` provide us with a better understanding
+# of the forecasts made by our time series regression model.
```
More than providing us with an "understanding" of the forecast, the quantile-based estimates provide an quantitive estimate of the uncertainty of the forecast.

rprkh commented:
Thank you for your reviews @ArturoAmorQ and @ogrisel. I will work on incorporating some of the smaller suggestions tomorrow  and I will add the bulk of the changes required for this PR over the weekend.

rprkh committed:
first pass at reviews

rprkh committed:
Merge branch 'doc_new_example_time_series' of https://github.com/rprkh/scikit-learn into doc_new_example_time_series

rprkh committed:
Merge branch 'main' into doc_new_example_time_series

rprkh commented on examples/applications/plot_time_series_lagged_features.py:
```diff
@@ -0,0 +1,433 @@
+)
+summary
+
+# %%
+# Taking a look at the total count of rented bikes for the first
+# week in our dataset.
+import matplotlib.pyplot as plt
+import seaborn as sns
+
+bike_count = df["count"]
+bike_count[: 7 * 24].plot(figsize=(15, 6))
+_ = plt.title("First week of bike rental count data")
+
+# %%
+# Next, let us look at the fraction of the rented fleet vs the number
+# number of hours.
+fig, ax = plt.subplots(figsize=(12, 4))
+bike_count.hist(bins=30, ax=ax)
+_ = ax.set(
+    xlabel="Fraction of rented fleet demand",
+    ylabel="Number of hours",
+)
```
Removed these plots from the tutorial, due to them already being available in another example.

rprkh committed:
second pass at reviews

rprkh committed:
fix df count error

rprkh committed:
bulk changes: third pass at reviews

rprkh committed:
newline error fix

rprkh commented on examples/applications/plot_time_series_lagged_features.py:
```diff
@@ -0,0 +1,433 @@
+
+This example demonstrates how pandas-engineered lagged features can be used
+for time series forecasting with
+:class:`sklearn.ensemble.HistGradientBoostingRegressor` on the Bike Sharing
```
Hey, out of curiosity what is the purpose of the `~` symbol? Both the links with and without `~` seem to work.

ArturoAmorQ replied:
During the rendering :class:`sklearn.ensemble.HistGradientBoostingRegressor` will yield a link with the full path, i.e. [sklearn.ensemble.HistGradientBoostingRegressor](https://scikit-learn.org/stable/modules/classes.html), whereas :class:`~sklearn.ensemble.HistGradientBoostingRegressor` yields a shorter path: [HistGradientBoostingRegressor](https://scikit-learn.org/stable/modules/classes.html).

rprkh commented:
Hey, @ArturoAmorQ @ogrisel I have addressed all the reviews marked as "resolved". 

As for the remaining ones I have a few questions:

> Instead of printing the results each time, maybe we can add the results to a dataframe in order to give a summary as a table similar to [the one on this example](https://scikit-learn.org/stable/auto_examples/ensemble/plot_gradient_boosting_quantile.html#analysis-of-the-error-metrics).
> 
> Please do let us know if you need help on this regard.

I took a look at the referenced link, and tried creating a similar table for this example. By making a change to the `evaluate` function, I came up with the following table:

![image](https://user-images.githubusercontent.com/75483881/218334728-bf31a1d1-d642-499d-9355-ef345ba934c2.png)

Should I include the above table to the example?
<br>
> There are no "perfect models" directly displayed on those plots. For instance, the "perfect" models for the 5%-percentile estimation would have 95% of the points above the diagonal and 5% below. Furthermore, those fractions should stay uniform when binning horizontally. But it's hard to see if our models are good or not in that respect using those plots. Maybe they are maybe not.

I'm not entirely sure what to change the "perfect models" label to in the plot.
<br>
> This two plots are quite interesting to explore the data! Following the logic of referencing the [Time-related feature engineering example](https://scikit-learn.org/stable/auto_examples/applications/plot_cyclical_feature_engineering.html) for data exploration, I suggest you open a PR to rather add this plots on that notebook.
> 
> Just keep in mind that "weekdays" is not the same as "days of the week" :)

As suggested, I will open a PR for this to improve the EDA for the other time series example  :)

rprkh committed:
fix rendering

ArturoAmorQ commented:
> Should I include the above table to the example?

Yes, please do so.

> I'm not entirely sure what to change the "perfect models" label to in the plot.

Maybe something like "best model". Then you would have to change text and legends in the plot accordingly.

> As suggested, I will open a PR for this to improve the EDA for the other time series example :)

Excellent! I will have a look at it.

rprkh committed:
include comparative table and minor fixes

rprkh committed:
fix array length error

rprkh commented:
Included the changes.

> As suggested, I will open a PR for this to improve the EDA for the other time series example :)
> > Excellent! I will have a look at it.

Also, opened #25602 to improve the EDA for the other time series example

ArturoAmorQ commented on examples/applications/plot_time_series_lagged_features.py:
```diff
@@ -0,0 +1,454 @@
+
+df_model_comparison = pd.DataFrame(dataframe_dict)
+df_model_comparison
+
```
Here we can add a paragraph mentioning that the minimal value for the Mean Pinball Loss is indeed achieved by the gbrt with the corresponding quantile. Similarly, the lowest RMSE is obtained by the "gbrt_mse" model, as expected. A bit less intuitive, the best MAPE corresponds to the "gbrt_poisson" model. This is due to the local nature of the Poisson Loss normalization.

ArturoAmorQ commented on examples/applications/plot_time_series_lagged_features.py:
```diff
@@ -0,0 +1,454 @@
+# Through this example we explored time series forecasting using lagged
+# features. We compared a naive regression (using the standardized
+# `train_test_split`) with a proper time series evaluation strategy using
+# `TimeSeriesSplit`. We observed that the model trained using `train_test_split`
```
Also remember that the problem is not the `train_test_split` itself, but setting `shuffle=True`.

ArturoAmorQ commented on examples/applications/plot_time_series_lagged_features.py:
```diff
@@ -0,0 +1,454 @@
+#   the 5% and 95% estimations.
+# - Finally, it is expected that 10% of the time, the actual demand does not lie
+#   between the 5% and 95% percentile estimates. On this test span, the actual
+#   demand seems to be higher, especially during the rush ours. It might reveal that
```
```suggestion
#   demand seems to be higher, especially during the rush hours. It might reveal that

```

ArturoAmorQ commented on examples/applications/plot_time_series_lagged_features.py:
```diff
@@ -0,0 +1,454 @@
+#   band.
+# - During the day, the blue band is much wider: the uncertainty grows, probably
+#   because of the variability of the weather that can have a very large impact,
+#   especially on a week-end day.
```
```suggestion
#   especially on week-ends.

```

ArturoAmorQ reviewed (commented):
Thanks for your dedication! Here is another batch of comments.

As a side note, notice you can commit the suggestions from github and then pull them to your local version. That makes reviews a bit easier ;)

rprkh committed:
remove "df" display

Co-authored-by: Arturo Amor <86408019+ArturoAmorQ@users.noreply.github.com>

rprkh committed:
remove boxplots

Co-authored-by: Arturo Amor <86408019+ArturoAmorQ@users.noreply.github.com>

rprkh committed:
add quotes to seasons

Co-authored-by: Arturo Amor <86408019+ArturoAmorQ@users.noreply.github.com>

rprkh committed:
unused import - lint

rprkh committed:
suggestions

rprkh committed:
links p1

rprkh committed:
suggestions and links

rprkh committed:
mapie and sktime

rprkh commented:
Added the changes.

> This links seem to be broken in the rendering (there is a blank space). Instead I recommend using [cross-references](https://scikit-learn.org/stable/developers/contributing.html#cross-referencing). You will have to add a line just above the "calibration of the confidence interval" section in plot_gradient_boosting_quantile.py.

I took a look at how to render links to other files. From what I understand there needs to be a reference `.. _calibration-section:` over here in `plot_gradient_boosting_quantile.py`
https://github.com/scikit-learn/scikit-learn/blob/d9cfe3f6b1c58dd253dc87cb676ce5171ff1f8a1/examples/ensemble/plot_gradient_boosting_quantile.py#L193-L194

And then the custom section needs to be referenced using :ref:`calibration of confidence intervals <calibration-section>`

If so, I will open a PR for adding the reference in `plot_gradient_boosting_quantile.py`

rprkh committed:
calibration link

rprkh committed:
Merge branch 'main' into doc_new_example_time_series

rprkh committed:
Merge branch 'main' into doc_new_example_time_series

ArturoAmorQ commented on examples/applications/plot_time_series_lagged_features.py:
```diff
@@ -0,0 +1,438 @@
+
+        return {
+            "mean_absolute_percentage_error": mean_absolute_percentage_error(y, y_pred),
+            "root_mean_squared_error": np.sqrt(mean_squared_error(y, y_pred)),
```
```suggestion
            "root_mean_squared_error": mean_squared_error(y, y_pred, squared=False),

```

rprkh committed:
update rmse

Co-authored-by: Arturo Amor <86408019+ArturoAmorQ@users.noreply.github.com>

rprkh committed:
lint

rprkh committed:
Merge branch 'main' into doc_new_example_time_series

rprkh committed:
ruff

ArturoAmorQ reviewed (commented):
I hope these suggestions make the code more readable.

rprkh committed:
suggestion1

Co-authored-by: Arturo Amor <86408019+ArturoAmorQ@users.noreply.github.com>

rprkh committed:
suggestion2

Co-authored-by: Arturo Amor <86408019+ArturoAmorQ@users.noreply.github.com>

rprkh committed:
Merge branch 'main' into doc_new_example_time_series

rprkh committed:
Merge branch 'main' into doc_new_example_time_series

rprkh committed:
use :math:

rprkh commented:
Hey, @ArturoAmorQ I have included your suggestions. I have also made use of `:math:` in my latest commit to improve the rendering for certain parts of the documentation and increase readability.

ArturoAmorQ committed:
Merge branch 'main' into doc_new_example_time_series

ArturoAmorQ commented on examples/applications/plot_time_series_lagged_features.py:
```diff
@@ -0,0 +1,394 @@
+    make_scorer,
+    mean_absolute_error,
+    mean_pinball_loss,
+    mean_squared_error,
```
```suggestion
    root_mean_squared_error,
```

ArturoAmorQ commented on examples/applications/plot_time_series_lagged_features.py:
```diff
@@ -0,0 +1,394 @@
+
+scoring = {
+    "MAPE": make_scorer(mean_absolute_percentage_error),
+    "RMSE": make_scorer(mean_squared_error, squared=False),
```
```suggestion
    "RMSE": make_scorer(root_mean_squared_error),

```

ArturoAmorQ reviewed (approved):
Sorry for still push some changes, the `squared` parameter of `mean_squared_error` got deprecated.

The current state of this PR still LGTM, but as it is not a minor change, I'd rather wait for the confirmation of any other developer.

ArturoAmorQ committed:
Apply suggestions from code review

ArturoAmorQ committed:
Merge branch 'main' into doc_new_example_time_series

ogrisel commented on examples/applications/plot_time_series_lagged_features.py:
```diff
@@ -0,0 +1,396 @@
+
+scoring = {
+    "MAPE": make_scorer(mean_absolute_percentage_error),
+    "RMSE": make_scorer(root_mean_squared_error),
+
+    "MAE": make_scorer(mean_absolute_error),
```
```suggestion
    "RMSE": make_scorer(root_mean_squared_error),
    "MAE": make_scorer(mean_absolute_error),
```

ogrisel committed:
Fix linter

ogrisel commented on examples/applications/plot_time_series_lagged_features.py:
```diff
@@ -0,0 +1,395 @@
+    for key, value in cv_results.items():
+        if key.startswith("test_"):
+            metric = key.split("test_")[1]
+            scores[metric].append(f"{value.mean():.3f} ± {value.std():.3f}")
```
The same change should be applied here. I thing you should factorize the post processing of the `cv_results` in a `consolidate_scores(cv_results, scores)` function to reuse twice.

ogrisel commented on examples/applications/plot_time_series_lagged_features.py:
```diff
@@ -0,0 +1,395 @@
+            scores[metric].append(f"{value.mean():.3f} ± {value.std():.3f}")
+
+scores = pd.DataFrame(scores)
+scores
```
You might need to adapt it a bit as follows:

```python
def highlight_min(col):
    if col.name == "loss":
        return ["" for v in col]
    num_values = [float(v.split(" ± ", 1)[0]) for v in col]
    v_min = min(num_values)
    return ["font-weight: bold" if v == v_min else "" for v in num_values]
```

Note: I have not tested this code myself.

rprkh replied:
I have modified the code after going through the `plot_gradient_boosting_quantile.py` example. I created a `mask` to highlight the minimum values in the dataframe and then applied that `mask` to generated the `styled_df_copy` dataframe  using `.style.apply(lambda x: mask, axis=None)`.

ogrisel commented on examples/applications/plot_time_series_lagged_features.py:
```diff
@@ -0,0 +1,395 @@
+# optimistic Mean Average Percentage Error (MAPE). The results
+# produced from the time-based split better represent the performance
+# of our time-series regression model. We also analyzed the predictive uncertainty
+# of our model via Quantile Regression. Predictions based on the 5th, 50th and
+# 95th percentile using `loss="quantile"` provide us with a quantitative estimate
```
```suggestion
# of our model via Quantile Regression. Predictions based on the 5th and
# 95th percentile using `loss="quantile"` provide us with a quantitative estimate
```

ogrisel commented on examples/applications/plot_time_series_lagged_features.py:
```diff
@@ -0,0 +1,395 @@
+plt.show()
+
+# %%
+# Conclusion
+# ------------------
```
```suggestion
# Conclusion
# ----------

```

ogrisel reviewed (commented):
Some more feedback but otherwise LGTM!

rprkh committed:
Update examples/applications/plot_time_series_lagged_features.py

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

rprkh committed:
Update examples/applications/plot_time_series_lagged_features.py

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

rprkh committed:
Update examples/applications/plot_time_series_lagged_features.py

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

rprkh committed:
header casing

rprkh commented:
I have added a few changes. I will work on including the remaining edits over the next few days.

rprkh committed:
fix conclusion

rprkh committed:
add consolidate_scores function

rprkh committed:
include code to highlight min values in the df

rprkh committed:
display styled df

rprkh committed:
Merge branch 'main' into doc_new_example_time_series

ArturoAmorQ commented on examples/applications/plot_time_series_lagged_features.py:
```diff
@@ -0,0 +1,433 @@
+
+
+# %%
+# The minimal value for the Mean Pinball Loss is achieved by the GBRT
+# with the corresponding quantile. Similarly, the lowest RMSE is obtained
+# by the "gbrt_mse" model, as expected. A bit less intuitive, the best MAPE
+# corresponds to the "gbrt_poisson" model. This is due to the local nature of
+# the Poisson Loss normalization.
```
At some point we got rid of those names for the models, so let's rephrase and be more accurate on the analysis:
```suggestion
# Even if the score distributions overlap due to the variance in the dataset, it
# is true that the average RMSE is lower when `loss="squared_error"`, whereas
# the average MAPE is lower when `loss="absolute_error"` as expected. That is
# also the case for the Mean Pinball Loss with the quantiles 5 and 95. The score
# corresponding to the 50 quantile loss is overlapping with the score obtained
# by minimizing other loss functions, which is also the case for the MAE.
```

ArturoAmorQ commented on examples/applications/plot_time_series_lagged_features.py:
```diff
@@ -0,0 +1,433 @@
+        X,
+        y,
+        cv=ts_cv,
+        scoring=scoring,
```
```suggestion
        scoring=scoring,
        n_jobs=2,
```

ArturoAmorQ commented on examples/applications/plot_time_series_lagged_features.py:
```diff
@@ -0,0 +1,433 @@
+        X,
+        y,
+        cv=ts_cv,
+        scoring=scoring,
```
```suggestion
        scoring=scoring,
        n_jobs=2,
```

ArturoAmorQ committed:
Apply suggestions from code review

ogrisel reviewed (approved):
This is great. Thanks very much everybody!
